bundles/org.openhab.binding.teleinfo/src/main/java/org/openhab/binding/teleinfo/internal/handler/TeleinfoThingHandlerFactory.java

```diff
@@ -0,0 +1,141 @@
+    protected void setSerialPortManager(final SerialPortManager serialPortManager) {
+        this.serialPortManager = serialPortManager;
+    }
+
+    protected void unsetSerialPortManager(final SerialPortManager serialPortManager) {
+        this.serialPortManager = null;
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (THING_TYPE_SERIAL_CONTROLLER.equals(thingTypeUID)) {
+            return new TeleinfoSerialControllerHandler((Bridge) thing, serialPortManager);
+        }
+
+        if (THING_BASE_CBEMM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoBaseCbemmElectricityMeterHandler(thing);
+        } else if (THING_HC_CBEMM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoHcCbemmElectricityMeterHandler(thing);
+        } else if (THING_TEMPO_CBEMM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoTempoCbemmElectricityMeterHandler(thing);
+        } else if (THING_EJP_CBEMM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoEjpCbemmElectricityMeterHandler(thing);
+        } else if (THING_BASE_CBEMM_EVO_ICC_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoBaseCbemmEvoIccElectricityMeterHandler(thing);
+        } else if (THING_HC_CBEMM_EVO_ICC_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoHcCbemmEvoIccElectricityMeterHandler(thing);
+        } else if (THING_TEMPO_CBEMM_EVO_ICC_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoTempoCbemmEvoIccElectricityMeterHandler(thing);
+        } else if (THING_EJP_CBEMM_EVO_ICC_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoEjpCbemmEvoIccElectricityMeterHandler(thing);
+        } else if (THING_BASE_CBETM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoBaseCbetmLongElectricityMeterHandler(thing);
+        } else if (THING_HC_CBETM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoHcCbetmLongElectricityMeterHandler(thing);
+        } else if (THING_TEMPO_CBETM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoTempoCbetmLongElectricityMeterHandler(thing);
+        } else if (THING_EJP_CBETM_ELECTRICITY_METER_TYPE_UID.equals(thing.getThingTypeUID())) {
+            return new TeleinfoEjpCbetmLongElectricityMeterHandler(thing);
+        } else {
+            throw new IllegalStateException("Teleinfo frame type not supported: " + thing.getThingTypeUID());
+        }
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
```

Comment:
This whole method can be removed here as the unregistering of the discovery sevice is handled by the `ThingHandlerService` now